README.md
# Resolution

[![NPM version](https://img.shields.io/npm/v/@unstoppabledomains/resolution.svg?style=flat)](https://www.npmjs.com/package/@unstoppabledomains/resolution)
![CI](https://github.com/unstoppabledomains/resolution/workflows/CI/badge.svg?branch=master)
[![Bundle Size Minified](https://img.shields.io/bundlephobia/min/@unstoppabledomains/resolution.svg)](https://bundlephobia.com/result?p=@unstoppabledomains/resolution)
[![Bundle Size Minified Zipped](https://img.shields.io/bundlephobia/minzip/@unstoppabledomains/resolution.svg)](https://bundlephobia.com/result?p=@unstoppabledomains/resolution)
[![Unstoppable Domains Documentation](https://img.shields.io/badge/Documentation-unstoppabledomains.com-blue)](https://docs.unstoppabledomains.com/)
[![Get help on Discord](https://img.shields.io/badge/Get%20help%20on-Discord-blueviolet)](https://discord.gg/b6ZVxSZ9Hn)

- [Installing Resolution](README.md#installing-resolution)
- [Using Resolution](README.md#using-resolution)
- [Default Ethereum Providers](README.md#default-ethereum-providers)
- [Error Handling](README.md#error-handling)
- [Free advertising for integrated apps](README.md#free-advertising-for-integrated-apps)

Resolution is a library for interacting with blockchain domain names. It can be
used to retrieve
[payment addresses](https://unstoppabledomains.com/features#Add-Crypto-Addresses),
IPFS hashes for
[decentralized websites](https://unstoppabledomains.com/features#Build-Website),
and GunDB usernames for
[decentralized chat](https://unstoppabledomains.com/chat).

Resolution is primarily built and maintained by
[Unstoppable Domains](https://unstoppabledomains.com/).

Resolution supports decentralized domains across three main zones:

- Zilliqa Name Service (ZNS)

  - `.zil`

- Unstoppable Name Service (UNS)
  - `.crypto`
  - `.nft`
  - `.blockchain`
  - `.bitcoin`
  - `.coin`
  - `.wallet`
  - `.888`
  - `.dao`
  - `.x`
  - ...

For more information, see our detailed
[API Referrence](https://unstoppabledomains.github.io/resolution/).

## Installing Resolution

Resolution can be installed with either `yarn` or `npm`.

```shell
yarn add @unstoppabledomains/resolution
```

```shell
npm install @unstoppabledomains/resolution --save
```

If you're interested in resolving domains via the command line, see our
[CLI section](#command-line-interface).

## Using Resolution

Create a new project.

```shell
mkdir resolution && cd $_
yarn init -y
yarn add @unstoppabledomains/resolution
```

### Look up a domain's crypto address

Create a new file in your project, `address.js`.

```javascript
const {default: Resolution} = require('@unstoppabledomains/resolution');
const resolution = new Resolution();

function resolve(domain, currency) {
  resolution
    .addr(domain, currency)
    .then((address) => console.log(domain, 'resolves to', address))
    .catch(console.error);
}

function resolveMultiChain(domain, currency, chain) {
  resolution
    .multiChainAddr(domain, currency, chain)
    .then((address) => console.log(domain, 'resolves to', address, version))
    .catch(console.error);
}

resolve('brad.crypto', 'ETH');
resolve('brad.zil', 'ZIL');
resolveMultiChain('brad.crypto', 'USDT', 'ERC20');
resolveMultiChain('brad.crypto', 'USDT', 'OMNI');
```

Execute the script.

```shell
$ node address.js
brad.crypto resolves to 0x8aaD44321A86b170879d7A244c1e8d360c99DdA8
brad.zil resolves to zil1yu5u4hegy9v3xgluweg4en54zm8f8auwxu0xxj
```

### Find the IPFS hash for a decentralized website

Create a new file in your project, `ipfs_hash.js`.

```javascript
const {default: Resolution} = require('@unstoppabledomains/resolution');
const resolution = new Resolution();

function resolveIpfsHash(domain) {
  resolution
    .ipfsHash(domain)
    .then((hash) =>
      console.log(
        `You can access this website via a public IPFS gateway: https://gateway.ipfs.io/ipfs/${hash}`,
      ),
    )
    .catch(console.error);
}

resolveIpfsHash('homecakes.crypto');
```

Execute the script.

```shell
$ node ipfs_hash.js
You can access this website via a public IPFS gateway: https://gateway.ipfs.io/ipfs/QmVJ26hBrwwNAPVmLavEFXDUunNDXeFSeMPmHuPxKe6dJv
```

### Find a custom record

Create a new file in your project, `custom-resolution.js`.

```javascript
const {default: Resolution} = require('@unstoppabledomains/resolution');
const resolution = new Resolution();

function resolveCustomRecord(domain, record) {
  resolution
    .records(domain, [record])
    .then((value) => console.log(`Domain ${domain} ${record} is: ${value}`))
    .catch(console.error);
}

resolveCustomRecord('homecakes.crypto', 'custom.record.value');
```

### Command Line Interface

CLI support was removed from the Resolution library starting from version 6.0. Please use the [standalone CLI tool](https://github.com/unstoppabledomains/resolution-cli).

## Default Ethereum Providers

Resolution provides zero-configuration experience by using built-in
production-ready [Infura](http://infura.io/) endpoint by default.  
Default Ethereum provider is free to use without restrictions and rate-limits
for `UNS` resolution.  
Default provider can be changed by changing constructor options
`new Resolution(options)` or by using one of the factory methods:

- `Resolution.infura()`
- `Resolution.fromWeb3Version1Provider()`
- `Resolution.fromEthersProvider()`
- etc.

To see all constructor options and factory methods check
[Unstoppable API reference](https://unstoppabledomains.github.io/resolution).

## Autoconfiguration of blockchain network

In some scenarios system might not be flexible enough to easy distinguish
between various Ethereum testnets on compile time. For this case Resolution
library provide a special async constructor which should be waited for
`await Resolution.autonetwork(options)`. This method makes a JSON RPC
"net_version" call to the provider to get the network id.

This method configures only Uns. Zns is supported only on Zilliqa mainnet which
is going to be used in any cases. You can provide a configured provider or a
blockchain url as in the following example:

```
await Resolution.autonetwork({
  uns: {provider},
});
```

## Error Handling

When resolution encounters an error it returns the error code instead of
stopping the process. Keep an eye out for return values like `RECORD_NOT_FOUND`.

## Development

Use these commands to set up a local development environment (**macOS Terminal**
or **Linux shell**).

1. Install `nvm`

   ```bash
   curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.36.0/install.sh | bash
   ```

2. Install concrete version of `node.js`

   ```bash
   nvm install 12.12.0
   ```

3. Install `yarn`
   ```bash
   npm install -g yarn
   ```
4. Clone the repository

   ```bash
   git clone https://github.com/unstoppabledomains/resolution.git
   cd resolution
   ```

5. Install dependencies
   ```bash
   yarn install
   ```

### Internal config

#### To update:

- Network config: `$ yarn network-config:pull`
- Resolver keys: `$ yarn resolver-keys:pull`
- Both configs: `$ yarn config:pull`

## Free advertising for integrated apps

Once your app has a working Unstoppable Domains integration,
[register it here](https://unstoppabledomains.com/app-submission). Registered
apps appear on the Unstoppable Domains
[homepage](https://unstoppabledomains.com/) and
[Applications](https://unstoppabledomains.com/apps) page — putting your app in
front of tens of thousands of potential customers per day.

Also, every week we select a newly-integrated app to feature in the Unstoppable
Update newsletter. This newsletter is delivered to straight into the inbox of
~100,000 crypto fanatics — all of whom could be new customers to grow your
business.

## Get help

[Join our discord community](https://discord.com/invite/b6ZVxSZ9Hn) and ask
questions.


HISTORY.md
## v7.1.2
- Removed unused code which caused build issues on some environments

## 7.1.0

- Throw `ResolutionErrorCode.InvalidDomainAddress` error if domain contains special characters
  - Domain name is being validated according to the following regular expression: `^[.a-z0-9-]+$`
## 7.0.0

- ENS support is completely removed
- Removes `bip44-constants`, `@ensdomains/address-encoder`, and `content-hash`
  package dependencies
- Methods that query ENS domains (.eth, .luxe, .xyz, .kred, .reverse) throw
  UnsupportedDomain
## 6.0.3
- Remove relative imports to avoid issues in bundlers. Restrict relative imports by adding eslint rule.

## 6.0.2
- `Resolution#owner` method doesn't throw an error in case of empty resolver

## 6.0.1
- Set correct polygon-mainnet provider if use default settings

## 6.0.0
### Breaking changes
- CLI tool was removed. Please use new binary CLI tool instead: [Github repo](https://github.com/unstoppabledomains/resolution-cli/releases).
- Constructor has changed. If you used `uns` configuration you have to specify parameters for L1 and L2 locations (Ethereum and Polygon mainnets).
- ENS support is deprecated and will be removed in future.
- Factory methods has changed. All of them requires provide L1 and L2 locations configuration for UNS. For example: `uns: {locations: {L1: ..., L2...}}`.
  - The list of affected methods:
    - `Resolution#infura`
    - `Resolution#fromResolutionProvider`
    - `Resolution#fromEthereumEip1193Provider`
    - `Resolution#fromWeb3Version0Provider`
    - `Resolution#fromWeb3Version1Provider`
    - `Resolution#fromEthersProvider`
- `Resolution#location` method is replaced by `Resolution#locations`.

### New methods and features
- 🎉 🎉 🎉 Add Polygon Layer 2 support!
- Add `Resolution#locations` method which will help to determine domains location (blockhain, networkId) and useful metadata like owner, resolver, registry addresses, provider url if possible.
  - Method returns:
    - Domain blockhain (ETH or MATIC)
    - Blockchain network id (numeric)
    - Owner address
    - Resolver address
    - Registry address
    - Provider URL if possible
      - Infura URL by default

## 5.0.2
- Replaces `node-fetch` with `cross-fetch` package

## 5.0.1
- Add `elliptic` package dependency in order to enable twitter verification support by default

## 5.0.0
### Breaking changes
 - Constructor has changed. If you used cns configurations rename "cns" it to "uns" instead.
    - For example: `new Resolution({sourceConfig: {uns: {...uns config}}})`
 - Method `Resolution#fromEip1193Provider` was renamed to `Resolution#fromEthereumEip1193Provider`  
 - Factory methods has changed. All of them requires `{uns: {... uns config}}` in parameters instead of `cns`.
   - The list of affected factory methods:
     - `Resolution#autoNetwork` 
     - `Resolution#infura` 
     - `Resolution#fromEthereumEip1193Provider` (former `Resolution#fromEip1193Provider`)
     - `Resolution#fromWeb3Version0Provider`
     - `Resolution#fromWeb3Version1Provider`
     - `Resolution#fromEthersProvider`
 - `Resolution#isSupportedDomain` method is now asynchronous

### New methods and features
- 🎉 🎉 🎉 Added support for new TLD's ( .888, .nft, .coin, .blockchain, .wallet, .x, .bitcoin, .dao )
- Typo fix: Rename `Eip1993Factories()` -> `Eip1193Factories()`. Old name is still available under the alias `Eip1993Factories`
- Introduced new method `Resolution#registryAddress -> Retrieves address of registry contract used for domain`
- Introduced new method `Resolution#unhash -> Retrieves the domain name from tokenId by parsing registry smart contract event logs`
- Introduced new method `Resolution#reverse -> This method is only for ens at the moment. Reverse the ens address to a ens registered domain name`
- Introduced new method `Resolution#tokenURI -> Retrieves the tokenURI from the registry smart contract`
- Introduced new method `Resolution#tokenURIMetadata -> etrieves the data from the endpoint provided by tokenURI from the registry smart contract`
- Introduced new factory method `Resolution#fromZilliqaProvider -> Creates a resolution instance with configured provider from Zilliqa provider`
- Introduced new factory method `Resolution#fromResolutionProvider -> Creates a resolution from Resolution compatitable provider`
- Introduced new factory method `Resolution#fromEthereumEip1193Provider -> Creates a resolution from EIP-1193 compatitable provider`
- Return ENS support
- Add custom network support for ENS

## 4.0.1 - 4.0.2

- No changes made. Version bump so that it would appear as latest version on
  NPM.

## 4.0.0

- Remove ENS support

## 3.0.0

- ENS support is disabled by default. To enable ENS support install additional
  packages:
  - `"bip44-constants": "^8.0.5"`
  - `"@ensdomains/address-encoder": ">= 0.1.x <= 0.2.x"`
  - `"content-hash": "^2.5.2"`
- If trying to resolve ENS domain and some package is missing the library throws
  `ConfigurationError`

## 2.1.0

- Introduce new factory method Resolution#autonetwork. This factory is
  asynchronious and allows to skip the network configuration for either ens or
  cns. This method is making a "net_version" call to the blockchain provider in
  order to configure itself.

## 2.0.0

- Remove deprecated methods
  - Resolution#address
  - Resolution#ipfsRedirect
  - Resolution#addressOrThrow
  - ResolutionErrorCode.UnspecifiedCurrency
- Simplify constructor
  - remove type Blockchain
  - remove type API
  - Introduced single config type SourceConfig

## 1.20.0

- Introduced Resolution#multiChainAddr(domain: string, ticker: string, chain:
  string) - More general method to fetch multi chain records.
- Deprecated Resolution#usdt method in favor of multiChainAddr
- Deprecated TickerVersion

## 1.19.0

- Update @ensdomains/address-encoder dependency to remove security audit issues
- Remove webpack dependency to remove security audit issues

## 1.18.0

- Use Infura Ethereum Provider by default

## 1.17.0

- Add support for TLOS

## 1.16.2


src/Ens.ts
import {default as ensInterface} from './contracts/ens/ens';
import {default as resolverInterface} from './contracts/ens/resolver';
import {
  BlockhainNetworkUrlMap,
  EnsSupportedNetwork,
  EthCoinIndex,
  hasProvider,
} from './types';
import {ResolutionError, ResolutionErrorCode} from './errors/resolutionError';
import EthereumContract from './contracts/EthereumContract';
import EnsNetworkMap from 'ethereum-ens-network-map';
import {
  EnsSource,
  Locations,
  NamingServiceName,
  Provider,
} from './types/publicTypes';
import {
  constructRecords,
  EthereumNetworksInverted,
  isNullAddress,
} from './utils';
import FetchProvider from './FetchProvider';
import {eip137Childhash, eip137Namehash} from './utils/namehash';
import {NamingService} from './NamingService';
import ConfigurationError, {
  ConfigurationErrorCode,
} from './errors/configurationError';
import {EthereumNetworks} from './utils';
import {requireOrFail} from './utils/requireOrFail';

/**
 * @internal
 */
export default class Ens extends NamingService {
  static readonly UrlMap: BlockhainNetworkUrlMap = {
    1: 'https://mainnet.infura.io/v3/d423cf2499584d7fbe171e33b42cfbee',
    3: 'https://ropsten.infura.io/v3/d423cf2499584d7fbe171e33b42cfbee',
    4: 'https://rinkeby.infura.io/v3/d423cf2499584d7fbe171e33b42cfbee',
  };

  readonly name = NamingServiceName.ENS;
  readonly network: number;
  readonly url: string | undefined;
  readonly provider: Provider;
  readonly readerContract: EthereumContract;

  constructor(
    source: EnsSource = {
      url: Ens.UrlMap[1],
      network: 'mainnet',
    },
  ) {
    super();
    this.checkNetworkConfig(source);
    this.network = EthereumNetworks[source.network];
    this.url = source['url'] || Ens.UrlMap[this.network];
    this.provider =
      source['provider'] || new FetchProvider(this.name, this.url!);
    const registryAddress =
      source['registryAddress'] || EnsNetworkMap[this.network];
    this.readerContract = new EthereumContract(
      ensInterface,
      registryAddress,
      this.provider,
    );
  }

  static async autoNetwork(
    config: {url: string} | {provider: Provider},
  ): Promise<Ens> {
    let provider: Provider;

    if (hasProvider(config)) {
      provider = config.provider;
    } else {
      if (!config.url) {
        throw new ConfigurationError(ConfigurationErrorCode.UnspecifiedUrl, {
          method: NamingServiceName.ENS,
        });
      }
      provider = FetchProvider.factory(NamingServiceName.ENS, config.url);
    }

    const networkId = (await provider.request({
      method: 'net_version',
    })) as number;
    const networkName = EthereumNetworksInverted[networkId];
    if (!networkName || !EnsSupportedNetwork.guard(networkName)) {
      throw new ConfigurationError(ConfigurationErrorCode.UnsupportedNetwork, {
        method: NamingServiceName.ENS,
      });
    }
    return new this({network: networkName, provider: provider});
  }

  serviceName(): NamingServiceName {
    return this.name;
  }

  namehash(domain: string): string {
    if (!this.checkSupportedDomain(domain)) {
      throw new ResolutionError(ResolutionErrorCode.UnsupportedDomain, {
        domain,
      });
    }
    return eip137Namehash(domain);
  }

  childhash(parentHash: string, label: string): string {
    return eip137Childhash(parentHash, label);
  }

  async isSupportedDomain(domain: string): Promise<boolean> {
    return this.checkSupportedDomain(domain);
  }

  async owner(domain: string): Promise<string> {
    const namehash = this.namehash(domain);
    return await this.callMethod(this.readerContract, 'owner', [namehash]);
  }

  async resolver(domain: string): Promise<string> {
    const nodeHash = this.namehash(domain);
    const resolverAddr = await this.callMethod(
      this.readerContract,
      'resolver',
      [nodeHash],
    );
    if (isNullAddress(resolverAddr)) {
      throw new ResolutionError(ResolutionErrorCode.UnspecifiedResolver);
    }
    return resolverAddr;
  }

  async record(domain: string, key: string): Promise<string> {
    const returnee = (await this.records(domain, [key]))[key];
    if (!returnee) {
      throw new ResolutionError(ResolutionErrorCode.RecordNotFound, {
        domain,
        recordName: key,
      });
    }
    return returnee;
  }

  async records(
    domain: string,
    keys: string[],
  ): Promise<Record<string, string>> {
    const values = await Promise.all(
      keys.map(async (key) => {
        if (key.startsWith('crypto.')) {
          const ticker = key.split('.')[1];
          return await this.addr(domain, ticker);
        }
        if (key === 'ipfs.html.value' || key === 'dweb.ipfs.hash') {
          return await this.getContentHash(domain);
        }
        const ensRecordName = this.fromUDRecordNameToENS(key);
        return await this.getTextRecord(domain, ensRecordName);
      }),
    );
    return constructRecords(keys, values);
  }

  async reverse(
    address: string,
    currencyTicker: string,
  ): Promise<string | null> {
    if (currencyTicker != 'ETH') {
      throw new Error(`Ens doesn't support any currency other than ETH`);
    }

    if (address.startsWith('0x')) {
      address = address.substr(2);
    }

    const reverseAddress = address + '.addr.reverse';
    const nodeHash = this.namehash(reverseAddress);
    const resolverAddress = await this.resolver(reverseAddress).catch(
      (err: ResolutionError) => {
        if (err.code === ResolutionErrorCode.UnspecifiedResolver) {
          return null;
        }
        throw err;
      },
    );

    if (isNullAddress(resolverAddress)) {
      return null;
    }

    const resolverContract = new EthereumContract(
      resolverInterface(resolverAddress, EthCoinIndex),
      resolverAddress,
      this.provider,
    );

    return await this.resolverCallToName(resolverContract, nodeHash);
  }

  async getTokenUri(tokenId: string): Promise<string> {
    throw new ResolutionError(ResolutionErrorCode.UnsupportedMethod, {
      method: NamingServiceName.ENS,
      methodName: 'getTokenUri',
    });
  }

  async isAvailable(domain: string): Promise<boolean> {
    return !(await this.isRegistered(domain));
  }

  async registryAddress(domain: string): Promise<string> {
    return this.readerContract.address;
  }

  async isRegistered(domain: string): Promise<boolean> {
    const address = await this.owner(domain);
    return !isNullAddress(address);
  }

  async getDomainFromTokenId(tokenId: string): Promise<string> {
    throw new ResolutionError(ResolutionErrorCode.UnsupportedMethod, {
      method: NamingServiceName.ENS,
      methodName: 'getDomainFromTokenId',
    });
  }

  locations(domains: string[]): Promise<Locations> {
    throw new ResolutionError(ResolutionErrorCode.UnsupportedMethod, {
      method: NamingServiceName.ENS,
      methodName: 'locations',
    });
  }

  /**
   * This was done to make automated tests more configurable
   */
  private resolverCallToName(resolverContract: EthereumContract, nodeHash) {
    return this.callMethod(resolverContract, 'name', [nodeHash]);
  }

  async twitter(domain: string): Promise<string> {
    throw new ResolutionError(ResolutionErrorCode.UnsupportedMethod, {
      domain,
      methodName: 'twitter',
    });
  }

  async allRecords(domain: string): Promise<Record<string, string>> {
    throw new ResolutionError(ResolutionErrorCode.UnsupportedMethod, {
      domain,
      methodName: 'allRecords',
    });
  }

  protected getCoinType(currencyTicker: string): string {
    const bip44constants = requireOrFail(
      'bip44-constants',
      'bip44-constants',
      '^8.0.5',
    );
    const formatsByCoinType = requireOrFail(
      '@ensdomains/address-encoder',
      '@ensdomains/address-encoder',
      '>= 0.1.x <= 0.2.x',
    ).formatsByCoinType;
    const coin = bip44constants.findIndex(
      (item) =>
        item[1] === currencyTicker.toUpperCase() ||
        item[2] === currencyTicker.toUpperCase(),
    );
    if (coin < 0 || !formatsByCoinType[coin]) {
      throw new ResolutionError(ResolutionErrorCode.UnsupportedCurrency, {
        currencyTicker,
      });
    }

    return coin.toString();
  }

  private fromUDRecordNameToENS(record: string): string {
    const mapper = {
      'ipfs.redirect_domain.value': 'url',
      'browser.redirect_url': 'url',
      'whois.email.value': 'email',
      'gundb.username.value': 'gundb_username',
      'gundb.public_key.value': 'gundb_public_key',
    };
    return mapper[record] || record;
  }

  private async addr(
    domain: string,
    currencyTicker: string,
  ): Promise<string | undefined> {
    const resolver = await this.resolver(domain).catch(
      (err: ResolutionError) => {
        if (err.code !== ResolutionErrorCode.UnspecifiedResolver) {
          throw err;
        }
      },
    );
    if (!resolver) {
      const owner = await this.owner(domain);
      if (isNullAddress(owner)) {
        throw new ResolutionError(ResolutionErrorCode.UnregisteredDomain, {
          domain,
        });
      }
      throw new ResolutionError(ResolutionErrorCode.UnspecifiedResolver, {
        domain,
      });
    }

    const cointType = this.getCoinType(currencyTicker.toUpperCase());
    return await this.fetchAddress(resolver, domain, cointType);
  }

  private async fetchAddress(
    resolver: string,
    domain: string,
    coinType: string,
  ): Promise<string | undefined> {
    const formatsByCoinType = requireOrFail(
      '@ensdomains/address-encoder',
      '@ensdomains/address-encoder',
      '>= 0.1.x <= 0.2.x',
    ).formatsByCoinType;
    const resolverContract = new EthereumContract(
      resolverInterface(resolver, coinType),
      resolver,
      this.provider,
    );
    const nodeHash = this.namehash(domain);
    const addr: string =
      coinType !== EthCoinIndex
        ? await this.callMethod(resolverContract, 'addr', [nodeHash, coinType])
        : await this.callMethod(resolverContract, 'addr', [nodeHash]);
    if (isNullAddress(addr)) {
      return undefined;
    }
    // eslint-disable-next-line no-undef
    const data = Buffer.from(addr.replace('0x', ''), 'hex');
    return formatsByCoinType[coinType].encoder(data);
  }

  private async getTextRecord(domain, key): Promise<string | undefined> {
    const nodeHash = this.namehash(domain);
    const resolver = await this.getResolverContract(domain);
    return await this.callMethod(resolver, 'text', [nodeHash, key]);
  }

  private async getContentHash(domain: string): Promise<string | undefined> {
    const contentHash = requireOrFail('content-hash', 'content-hash', '^2.5.2');
    const nodeHash = this.namehash(domain);
    const resolverContract = await this.getResolverContract(domain);
    const contentHashEncoded = await this.callMethod(
      resolverContract,
      'contenthash',
      [nodeHash],
    );
    const codec = contentHash.getCodec(contentHashEncoded);
    if (codec !== 'ipfs-ns') {
      return undefined;
    }
    return contentHash.decode(contentHashEncoded);
  }

  private async getResolverContract(
    domain: string,
    coinType?: string,
  ): Promise<EthereumContract> {
    const resolverAddress = await this.resolver(domain);
    return new EthereumContract(
      resolverInterface(resolverAddress, coinType),
      resolverAddress,
      this.provider,
    );
  }

  private async callMethod(
    contract: EthereumContract,
    method: string,
    params: (string | string[])[],
  ): Promise<any> {
    const result = await contract.call(method, params);
    return result[0];
  }

  private checkSupportedDomain(domain: string): boolean {
    return (
      domain === 'eth' ||
      (/^[^-]*[^-]*\.(eth|luxe|xyz|kred|addr\.reverse)$/.test(domain) &&
        domain.split('.').every((v) => !!v.length))
    );
  }

  private checkNetworkConfig(source: EnsSource): void {
    if (!source.network) {
      throw new ConfigurationError(ConfigurationErrorCode.UnsupportedNetwork, {
        method: this.name,
      });
    }
    if (!EnsSupportedNetwork.guard(source.network)) {
      this.checkCustomNetworkConfig(source);
    }
  }

  private checkCustomNetworkConfig(source: EnsSource): void {
    if (!this.isValidRegistryAddress(source.registryAddress)) {
      throw new ConfigurationError(
        ConfigurationErrorCode.InvalidConfigurationField,
        {
          method: this.name,
          field: 'registryAddress',
        },
      );
    }
    if (!source['url'] && !source['provider']) {
      throw new ConfigurationError(
        ConfigurationErrorCode.CustomNetworkConfigMissing,
        {
          method: this.name,
          config: 'url or provider',
        },
      );
    }
  }

  private isValidRegistryAddress(address?: string): boolean {
    if (!address) {
      throw new ConfigurationError(
        ConfigurationErrorCode.CustomNetworkConfigMissing,
        {
          method: this.name,
          config: 'registryAddress',
        },
      );
    }
    const ethLikePattern = new RegExp('^0x[a-fA-F0-9]{40}$');
    return ethLikePattern.test(address);
  }
}
